Use XBlocks report data formatter to build problem responses

xitij2000 opened the pull request:
This PR build on #17776 and #17948  by using ``generate_report_data`` to get report data for blocks if the block provides it. 

**Project details**:
This PR is part of the "Improved Problem Response Report" project, funded by Australian National University (ANU) and developed by OpenCraft

**JIRA tickets**: TBD

**Dependencies**: #17776 and #17948 

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:
TBD

**Reviewers**
- [ ] @symbolist
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
xqueue_source_repo: https://github.com/edx/xqueue
xqueue_version: master
CERTS_REPO: https://github.com/edx/edx-certificates
certs_version: master
```
**Note**: The above settings are not relevant to the actual content of the PR

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [OSPR-2369](https://openedx.atlassian.net/browse/OSPR-2369) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

mduboseedx commented:
@xitij2000 Just to check in, is this still a work in progress or is it ready for review?

xitij2000 commented:
@mduboseedx Yes. This is a little dependent on #17948 so I'm trying to coordinate with that. Will ping when it's ready for review, which should be this week.

bradenmacdonald commented on lms/djangoapps/instructor_task/tasks_helper/grades.py:
```diff
@@ -548,6 +552,90 @@ def _graded_scorable_blocks_to_header(cls, course):
+            if hasattr(block, 'generate_report_data'):
+                generate_report_data = block.generate_report_data
+            elif hasattr(block, 'module_class') and hasattr(block.module_class, 'generate_report_data'):
+                generate_report_data = block.module_class.generate_report_data
```
@xitij2000 I don't think `generate_report_data` should be a staticmethod, because then it can't access the XBlock's `Scope.settings` and `Scope.content` fields.

How about using an approach like:

```python
from xmodule.modulestore.django import modulestore
from courseware.user_state_client import DjangoXBlockUserStateClient
from opaque_keys.edx.keys import UsageKey
user_state_client = DjangoXBlockUserStateClient()

block = modulestore().get_item(block_key)
user_state_iterator = user_state_client.iter_all_for_block(block_key)
block.generate_report_data(user_state_iterator)
```

You can try this with both an XBlock and an XModule using this example code I whipped up:
* Drag and Drop: https://github.com/open-craft/xblock-drag-and-drop-v2/commit/953c30c31983b085a3ea0110807e4a7a43e42dea
* Capa: https://github.com/open-craft/edx-platform/commit/a8136e2cf85f65b06927adc597ed378e5cbdbd04

xitij2000 replied:
@bradenmacdonald  Ah! I think if the method is on the descriptor then there is no problem. My issue was that I only had access to the descriptor or the module class, and I didn't want to code the differnt initialization paths for differnt kinds of blocks in the report generatioon code, so I recommended the static approach.

bradenmacdonald replied:
@xitij2000 Yep. Descriptors are not specific to a student, while modules are, so it makes sense for this code to be on a Descriptor only. Newer XBlocks (like drag and drop) aren't split into descriptor vs. module, so for pure XBlocks it doesn't matter. But for older XModules like capa, this sort of code should be on the descriptor if we must choose.

symbolist commented on lms/djangoapps/instructor_task/tasks_helper/grades.py:
```diff
@@ -548,6 +553,92 @@ def _graded_scorable_blocks_to_header(cls, course):
+            user_id (int): The user id for the user generating the report
+            course_id (CourseKey): The ``CourseKey`` for the course whose report
+                is being generated
+            problem_location (str): The generated report will include this
```
Would be a lot more consistent to call them `course_key` and `usage_key_string`.

symbolist commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -462,25 +462,99 @@ def test_team_deleted(self):
         with patch('lms.djangoapps.instructor_task.tasks_helper.runner._get_current_task'):
-            with patch('lms.djangoapps.instructor_task.tasks_helper.grades.list_problem_responses') as patched_data_source:
+            with patch('lms.djangoapps.instructor_task.tasks_helper.grades'
+                       '.ProblemResponses._build_student_data') as patched_data_source:
```
Nit: The convention is use the name `mock_<original_name>`.

symbolist commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -462,25 +462,99 @@ def test_team_deleted(self):
+            'username': 'student',
+            'location': 'test_course > Section > Subsection > Problem1',
+            'block_key': 'i4x://edx/1.23x/problem/Problem1',
+            'title': 'Problem1',
```
`state` should also be checked here.

symbolist commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -462,25 +462,99 @@ def test_team_deleted(self):
+                'username': 'student',
+                'location': 'test_course > Section > Subsection > Problem1',
+                'block_key': 'i4x://edx/1.23x/problem/Problem1',
+                'title': 'Problem1',
```
`state` should also be checked here.

xitij2000 commented:
@symbolist  I've addressed your comments. 
I had to adapt the tests to work without the ``generate_report_data`` method on CapaDescriptor. So that will need to be adjusted based on when this lands.

symbolist commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -462,25 +462,101 @@ def test_team_deleted(self):
+            'block_key': 'i4x://edx/1.23x/problem/Problem1',
+            'title': 'Problem1',
+        }, student_data[0])
+        self.assertIn('state', student_data[0])
```
Can we not verify the exact value of `state`? If not, another thing that can be checked is if `list_problem_responses()` is called. The point being, if someone were add `generate_report_data` to whatever block type is being used in this test, the test will continue to pass even though it will not be taking the code path it is taking right now.

symbolist replied:
On second thoughts, since the other asserts in these two tests are the same, you should assert that `list_problem_responses()` is called in this test and not in the next test.

Here is an example of how to set it up: https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/tests/test_email.py#L529-L533

symbolist replied:
Sounds good.

symbolist commented on lms/djangoapps/instructor_task/tests/test_tasks_helper.py:
```diff
@@ -462,25 +462,101 @@ def test_team_deleted(self):
+        self.assertIn('state', student_data[0])
+
+    @patch('xmodule.capa_module.CapaDescriptor.generate_report_data', create=True)
+    def test_build_student_data_generate(self, mock_capa_descriptor):
```
Also the following names would make it easier for future readers to understand what the tests are doing.

`test_build_student_data_for_block_without_generate_report_data()`
`test_build_student_data_for_block_with_generate_report_data()`

Also apologies if I wasn't clearer before; `mock_capa_descriptor` should be `mock_generate_report_data` since it is the method which is going to be a mock and not the class.

symbolist commented:
@xitij2000 I have given this another pass. The tests look better now. But they need one more assert each to verify that they are checking the right code path.

edx-webhook commented:
@xitij2000 Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18013&entry.1484655318=https://github.com/xitij2000&entry.752974735=2018-04-20+20:59&entry.1917517419=2018-05-16+22:33&entry.2133058324=No) so we can improve your experience in the future.
